src/constraintSolving.js

```diff
@@ -33,26 +33,74 @@ class EqConstraint {
   }
 
   replace(typeDictionary) {
-    return new EqConstraint(this.a.replacing(typeDictionary), this.b.replacing(typeDictionary))
+    return this.a.replacing(typeDictionary).eqConstraints(this.b.replacing(typeDictionary))
   }
 
   solve() {
-    if(this.a.isSameType(this.b)) {
-      return {}
-    } else if (this.a.isVarType() && this.b.includes(this.a) || this.b.isVarType() && this.b.includes(this.b)) {
-      return { error: "Impossible recursive type" }
-    } else if (this.a.isVarType()) {
-      return { [this.a.typeVariableName]: this.b }
-    } else if (this.b.isVarType()) {
-      return { [this.b.typeVariableName]: this.a }
-    } else if (this.a.isFunctionType() && this.b.isFunctionType()) {
-      return solveConstraints({constraints: this.a.eqConstraints(this.b)}).typeDictionary
-    } else {
-      return { error: "Type Error" }
+    try {
+      return this.doSolve()
+    } catch (e) {
+      return { error: e }
     }
   }
 
+  doSolve() {
+    throw "Subclass responsibility"
+  }
+
   toString() {
     return this.a.toString() + " ~ " + this.b.toString()
   }
-}
\ No newline at end of file
+}
+
+class SimpleEqConstraint extends EqConstraint {
+  constructor(simpleType, otherType) {
+    super(simpleType, otherType)
+    this.simpleType = simpleType
+    this.otherType = otherType
+  }
+
+  doSolve() {
+    try {
+      return this.otherType.restrictToSimple(this.simpleType)
+    } catch (e) {
+      return { error: e }
+    }
+  }
+}
+
+class CompositeEqConstraint extends EqConstraint {
```

Comment:
I'm not sure about adding new constraint types, I liked the constraint solving being kind of straightforward and having more of the heavy work (composite functions returning several constraints) being done previously.
Even though solving some constraints that involve composite types can return in more constraints down the road :thinking: , but still I think it's getting a bit more complex now

Reply:
Esto lo hice exclusivamente para sacar los ifs que estaban antes, fue medio mecánico (if -> polimorfismo), pero quedaron cosas que no me terminan de cerrar (ni antes ni ahora).
Lo que me incentivó y guió un poco a este resultado fue:
1. Me daba la sensación que se perdía info al momento de instanciar las Constraints: éstas nacen de los tipos, pero igual se hacía un type check al omento de resolverlas. Entonces pensé en dejar de alguna manera atada la constraint al tipo de donde se instancia. Si la constraint es para un tipo función, entonces debería checkar "los tipos función". 
2. La razón más fuerte por la que metí el refactor es que antes se hacía un doble dispatch entre los tipos para detectar errores "estructurales" al momento de crear las constraints (con esa constraint de error). Ese doble dispatch se iba a poner fulero al agregar los nuevos tipos de datos, así que lo quería sacar. Además quería separar la instanciación de constraints de su resolución / detección de errores. Eso podría permitir en un futuro reportar errores más piolas.

- Después me quedó picando eso de que en los tipos funciones (compuestos en general) se vuelve a llamar a `solveConstraints` y no manejar todo en la misma recursividad. Por tu comentario entiendo que lo hiciste apropósito, pero pensaba si no se podría resolver todo en el algoritmo general, haciendo que al resolver una constraint se obtuvieran nuevas (además del mapa que actualmente devuelven) ????